csrc/host_ir/lower.h

```diff
@@ -22,11 +22,11 @@ class HostIrLower {
   static bool canLower(Expr* expr, bool ignore_inner_resharding = false);
 
   // Lower a sharded Expr into a series of Communication.
-  static std::vector<Expr*> lower(Expr* c);
+  static std::vector<Expr*> lower(Expr* c, DeviceIdxType my_device_index);
 
   static std::unique_ptr<hir::HostIrContainer> lower(
       std::unique_ptr<Fusion> fusion,
-      int64_t my_device_index);
+      DeviceIdxType my_device_index);
```

Comment:
Question: will lowering have to take the current GPU ID? I understand 2D meshes lead to different `team`s so host IR as is has to be different. An alternative would be for each Communication to take `teams` (a vector of vectors which include all GPUs) rather than `team` (the GPUs in the team that the current GPU is in). Only at runtime, each device looks up which team in `teams` it's in. This way, all ranks have the same host IR (except for pipeline parallelism) so we can distribute the lowering on all ranks for compilation speed. 

cc @samnordmann

Reply:
Lowering does not have to take the GPU IDs, but more a matter of where to shift to overhead from lookup and what parallelism strategies you expect to support.
If you expect `teams` to be dynamic throughout training, basically dynamic meshes for each run, obviously you only know the `team` at runtime. However, if we expect one training job to keep the mesh static, then moving the creation into compile time and push any overheads into a one time cost.

The vector of teams is interesting, it does save lowering time since it pushes that cost into run time. Run time lookup is probably not an issue since the mesh's will only be within a node so O10s device.

Reply:
> Question: will lowering have to take the current GPU ID? I understand 2D meshes lead to different `team`s so host IR as is has to be different. An alternative would be for each Communication to take `teams`

Yes, sounds good. An alternative would be that the Communication only takes one team, but `Team` inherits from Val* and is bound to concrete value through `expr_evalutator_`. That Team could be defined through an Expr* that indicates taking the mesh's slice over some axis at my_device_id. This way, the `Communication` IR is symmetric across all ranks, and it covers both static and dynamic case.